feat: add `lapack/base/dlasrt` (#12354)
stdlib-js/stdlib: lib/node_modules/@stdlib/lapack/base/dlasrt/README.md

<!--

@license Apache-2.0

Copyright (c) 2026 The Stdlib Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

-->

# dlasrt

> > Sort an array of doubles in increasing or decreasing order using quicksort, with insertion sort for small partitions (size <= 20).

Reviewer: ```suggestion
> Sort an array of double-precision floating-point numbers in increasing or decreasing order using quicksort, with insertion sort for small partitions (size <= 20).
```

<!--

@license Apache-2.0

Copyright (c) 2026 The Stdlib Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

-->

# dlasrt

> Sort an array of double-precision floating-point numbers in increasing or decreasing order using quicksort, with insertion sort for small partitions (size <= 20).

<section class="usage">

## Usage

```javascript
var dlasrt = require( '@stdlib/lapack/base/dlasrt' );
```

#### dlasrt( sortOrder, N, D )

Sorts a double-precision floating-point strided array in increasing or decreasing order.

```javascript
var Float64Array = require( '@stdlib/array/float64' );

var D = new Float64Array( [ 5.0, 7.0, 9.0, 3.0, 11.0 ] );

var out = dlasrt( 'increasing', 5, D );
// out => 0

> // D => <Float64Array>[ 3, 5, 7, 9, 11 ]

Reviewer 2: ```suggestion
// D => <Float64Array>[ 3.0, 5.0, 7.0, 9.0, 11.0 ]

var bool = ( out === 0 );
// returns true
```

Decimals.

```

The function has the following parameters:

> -   **sortOrder**: sort direction, where `'increasing'` corresponds to increasing order and `'decreasing'` corresponds to decreasing order.

Reviewer 2: ```suggestion
-   **sortOrder**: sort order. If first letter of `sortOrder` is `'i'` (uppercase or lowercase), the input array is sorted in **increasing** order. If first letter of `sortOrder` is `'d'` (uppercase or lowercase), the input array is sorted in **decreasing** order. Otherwise, the input array is left unchanged.
```

Use our existing documentation as a reference. This should be a theme: we value consistency.

-   **N**: number of elements to sort.
-   **D**: [`Float64Array`][@stdlib/array/float64] array to sort in-place.

> The function returns a status code. If successful, the function returns `0`. Otherwise, the function returns `-i`, where `i` is the index of the argument having an invalid value.

Reviewer 2: Move this to a note in the notes section below.

Note that indexing is relative to the first index. To introduce an offset, use [`typed array`][mdn-typed-array] views.

<!-- eslint-disable stdlib/capitalized-comments -->

```javascript
var Float64Array = require( '@stdlib/array/float64' );

// Initial arrays...
var D0 = new Float64Array( [ 9999.0, 5.0, 7.0, 9.0, 3.0, 11.0 ] );

> var D = new Float64Array( D0.buffer, D0.BYTES_PER_ELEMENT*1 ); // start at 2nd element

Reviewer 2: ```suggestion
// Initial array:
var D0 = new Float64Array( [ 9999.0, 5.0, 7.0, 9.0, 3.0, 11.0 ] );

// Create an offset view:
var D = new Float64Array( D0.buffer, D0.BYTES_PER_ELEMENT*1 ); // start at 2nd element
```

var out = dlasrt( 'increasing', 5, D );
// out => 0

> // D0 => <Float64Array>[ 9999, 3, 5, 7, 9, 11 ]

Reviewer 2: ```suggestion
dlasrt( 'increasing', 5, D );
// D0 => <Float64Array>[ 9999.0, 3.0, 5.0, 7.0, 9.0, 11.0 ]
```

```

> #### dlasrt.ndarray( sortOrder, N, D, sD, oD, stack, sStack, oStack )

Reviewer 2: ```suggestion
#### dlasrt.ndarray( sortOrder, N, D, strideD, offsetD, stack, strideS, offsetS )
```

Reviewer 2: Follow conventions.

> Sorts a double-precision floating-point strided array in increasing or decreasing order and using alternative indexing semantics.

Reviewer 2: ```suggestion
Sorts a double-precision floating-point strided array in increasing or decreasing order using alternative indexing semantics.
```

```javascript
var Float64Array = require( '@stdlib/array/float64' );

var D = new Float64Array( [ 5.0, 7.0, 9.0, 3.0, 11.0 ] );
var stack = new Float64Array( 64 );

dlasrt.ndarray( 'increasing', 5, D, 1, 0, stack, 1, 0 );

> // D => <Float64Array>[ 3, 5, 7, 9, 11 ]

Reviewer 2: ```suggestion
// D => <Float64Array>[ 3.0, 5.0, 7.0, 9.0, 11.0 ]
```

```

The function has the following additional parameters:

-   **sD**: stride length for `D`.
-   **oD**: starting index for `D`.
-   **stack**: stack array.
-   **sStack**: stride length for `stack`.

> -   **oStack**: starting index for `stack`.

Reviewer 2: ```suggestion
-   **strideD**: stride length for `D`.
-   **offsetD**: starting index for `D`.
-   **stack**: stack array.
-   **strideS**: stride length for `stack`.
-   **offsetS**: starting index for `stack`.
```

While [`typed array`][mdn-typed-array] views mandate a view offset based on the underlying buffer, the offset parameters support indexing semantics based on starting indices. For example,

<!-- eslint-disable max-len -->

```javascript
var Float64Array = require( '@stdlib/array/float64' );

var D = new Float64Array( [ 9999.0, 5.0, 7.0, 9.0, 3.0, 11.0 ] );
var stack = new Float64Array( 65 );

dlasrt.ndarray( 'increasing', 5, D, 1, 1, stack, 1, 1 );

> // D => <Float64Array>[ 9999, 3, 5, 7, 9, 11 ]

Reviewer 2: ```suggestion
// D => <Float64Array>[ 9999.0, 3.0, 5.0, 7.0, 9.0, 11.0 ]
```

```

</section>

<!-- /.usage -->

<section class="notes">

## Notes

-   `dlasrt()` corresponds to the [LAPACK][LAPACK] routine [`dlasrt`][lapack-dlasrt].

</section>

<!-- /.notes -->

<section class="examples">

## Examples

<!-- eslint no-undef: "error" -->

<!-- eslint-disable max-len -->

```javascript
var Float64Array = require( '@stdlib/array/float64' );
var dlasrt = require( '@stdlib/lapack/base/dlasrt' );

var D = new Float64Array( [ 12.0, -3.0, 7.5, 0.0, 19.0, -8.0, 4.0, 4.0, 15.0, -1.0, 2.2, 9.0 ] );

> 
> dlasrt( 'increasing', 12, D );
> console.log( D );
> 
> // Redefine the `D` array as it gets updated.
> D = new Float64Array( [ 12.0, -3.0, 7.5, 0.0, 19.0, -8.0, 4.0, 4.0, 15.0, -1.0, 2.2, 9.0 ] );
> var stack = new Float64Array( 64 );
> 
> dlasrt.ndarray( 'D', 12, D, 1, 0, stack, 1, 0 );
> console.log( D );

Reviewer 2: ```suggestion
var D = new Float64Array( [ 12.0, -3.0, 7.5, 0.0, 19.0, -8.0, 4.0, 4.0, 15.0, -1.0, 2.2, 9.0 ] );

dlasrt( 'increasing', 12, D );
console.log( D );

D = new Float64Array( [ 12.0, -3.0, 7.5, 0.0, 19.0, -8.0, 4.0, 4.0, 15.0, -1.0, 2.2, 9.0 ] );
var stack = new Float64Array( 64 );

dlasrt.ndarray( 'D', 12, D, 1, 0, stack, 1, 0 );
console.log( D );
```

Note that indexing is relative to the first index. To introduce an offset, use [`typed array`][mdn-typed-array] views.

<!-- eslint-disable stdlib/capitalized-comments -->

```javascript
var Float64Array = require( '@stdlib/array/float64' );

// Initial array:
var D0 = new Float64Array( [ 9999.0, 5.0, 7.0, 9.0, 3.0, 11.0 ] );

// Create an offset view:
var D = new Float64Array( D0.buffer, D0.BYTES_PER_ELEMENT*1 ); // start at 2nd element

dlasrt( 'increasing', 5, D );
// D0 => <Float64Array>[ 9999.0, 3.0, 5.0, 7.0, 9.0, 11.0 ]
```

#### dlasrt.ndarray( sortOrder, N, D, strideD, offsetD, stack, strideS, offsetS )

Sorts a double-precision floating-point strided array in increasing or decreasing order using alternative indexing semantics.

```javascript
var Float64Array = require( '@stdlib/array/float64' );

var D = new Float64Array( [ 5.0, 7.0, 9.0, 3.0, 11.0 ] );
var stack = new Float64Array( 64 );

dlasrt.ndarray( 'increasing', 5, D, 1, 0, stack, 1, 0 );
// D => <Float64Array>[ 3.0, 5.0, 7.0, 9.0, 11.0 ]
```

The function has the following additional parameters:

-   **strideD**: stride length for `D`.
-   **offsetD**: starting index for `D`.

> -   **stack**: stack array.

Reviewer 2: ```suggestion
-   **stack**: workspace array. Must have at least `64` indexed elements.
```
